compiler/expressions.go

```diff
@@ -475,9 +475,21 @@ func (fc *funcContext) translateExpr(expr ast.Expr) *expression {
 			}
 			key := fmt.Sprintf("%s.keyFor(%s)", fc.typeName(t.Key()), fc.translateImplicitConversion(e.Index, t.Key()))
 			if _, isTuple := exprType.(*types.Tuple); isTuple {
-				return fc.formatExpr(`(%1s = %2e[%3s], %1s !== undefined ? [%1s.v, true] : [%4e, false])`, fc.newVariable("_entry"), e.X, key, fc.zeroValue(t.Elem()))
-			}
-			return fc.formatExpr(`(%1s = %2e[%3s], %1s !== undefined ? %1s.v : %4e)`, fc.newVariable("_entry"), e.X, key, fc.zeroValue(t.Elem()))
+				return fc.formatExpr(
+					`(%1s = typeof %2e.get === "function" ? %2e.get(%3s) : undefined, %1s !== undefined ? [%1s.v, true] : [%4e, false])`,
```

Comment:
Here and below: let's make the generated code a bit more compact by adding a helper function in prelude:

```js
$mapIndex = function(m, key) {
  return typeof m.get === "function" ? m.get(key) : undefined;
};
```

```
(%1s = $mapIndex(%2e, %3s), %1s !== undefined ? [%1s.v, true] : [%4e, false])
```

This will also clarify operator precedence in this expression.